File: ios/Sources/Mia21/Services/ConversationService.swift
//
//  ConversationService.swift
//  Mia21
//
//  Created on November 17, 2025.
//  Copyright © 2025 Mia21. All rights reserved.
//
//  Description:
//  Service layer for conversation history operations.
//  Handles listing, retrieving, and deleting conversations.
//

import Foundation

// MARK: - Conversation Service Protocol

protocol ConversationServiceProtocol {
  func listConversations(userId: String, spaceId: String?, limit: Int) async throws -> [ConversationSummary]
  func getConversation(conversationId: String) async throws -> ConversationDetail
  func deleteConversation(conversationId: String) async throws -> DeleteConversationResponse
  func renameConversation(conversationId: String, title: String) async throws -> RenameConversationResponse
  func deleteUserData(userId: String) async throws -> DeleteUserDataResponse
}

// MARK: - Conversation Service Implementation

final class ConversationService: ConversationServiceProtocol {

  // MARK: - Properties

  private let apiClient: APIClientProtocol

  // MARK: - Initialization

  init(apiClient: APIClientProtocol) {
    self.apiClient = apiClient
  }

  // MARK: - Public Methods

  func listConversations(userId: String, spaceId: String?, limit: Int) async throws -> [ConversationSummary] {
    logInfo("Listing conversations for user: \(userId), space: \(spaceId ?? "all"), limit: \(limit)")

    var queryParams: [String: String] = [
      "user_id": userId,
      "limit": String(limit)
    ]

    if let spaceId = spaceId {
      queryParams["space_id"] = spaceId
    }

    // Build query string
    let queryString = queryParams
      .map { "\($0.key)=\($0.value)" }
      .joined(separator: "&")

    let endpoint = APIEndpoint(
      path: "/conversations?\(queryString)",
      method: .get
    )

    let conversations: [ConversationSummary] = try await apiClient.performRequest(endpoint)

    logInfo("Retrieved \(conversations.count) conversations")
    
    // Log conversation details
    for (index, conv) in conversations.prefix(5).enumerated() {
      logDebug("  [\(index)] ID: \(conv.id), Space: \(conv.spaceId), Bot: \(conv.botId ?? "nil"), Messages: \(conv.messageCount), Title: \(conv.title ?? "nil")")
    }
    if conversations.count > 5 {
      logDebug("  ... and \(conversations.count - 5) more conversations")
    }
    
    return conversations
  }

  func getConversation(conversationId: String) async throws -> ConversationDetail {
    logInfo("Getting conversation: \(conversationId)")

    let endpoint = APIEndpoint(
      path: "/conversations/\(conversationId)",
      method: .get
    )

    let conversation: ConversationDetail = try await apiClient.performRequest(endpoint)

    logInfo("Retrieved conversation with \(conversation.messages.count) messages")
    logDebug("  Conversation ID: \(conversation.id)")
    logDebug("  Space: \(conversation.spaceId), Bot: \(conversation.botId ?? "nil")")
    logDebug("  Status: \(conversation.status)")
    logDebug("  Created: \(conversation.createdAt), Updated: \(conversation.updatedAt)")
    logDebug("  Message count: \(conversation.messages.count)")
    
    // Log first few messages
    for (index, message) in conversation.messages.prefix(3).enumerated() {
      let preview = message.content.prefix(50)
      logDebug("    [\(index)] \(message.role): \(preview)\(message.content.count > 50 ? "..." : "")")
    }
    if conversation.messages.count > 3 {
      logDebug("    ... and \(conversation.messages.count - 3) more messages")
    }
    
    return conversation
  }

  func deleteConversation(conversationId: String) async throws -> DeleteConversationResponse {
    logInfo("Deleting conversation: \(conversationId)")

    let endpoint = APIEndpoint(
      path: "/conversations/\(conversationId)",
      method: .delete
    )

    let response: DeleteConversationResponse = try await apiClient.performRequest(endpoint)

    logInfo("Successfully deleted conversation: \(conversationId)")
    logDebug("  Response: \(response.message)")
    
    return response
  }
  
  func renameConversation(conversationId: String, title: String) async throws -> RenameConversationResponse {
    logInfo("Renaming conversation: \(conversationId) to '\(title)'")

    let body: [String: Any] = ["title": title]
    
    let endpoint = APIEndpoint(
      path: "/conversations/\(conversationId)",
      method: .patch,
      body: body
    )

    let response: RenameConversationResponse = try await apiClient.performRequest(endpoint)

    logInfo("Successfully renamed conversation: \(conversationId)")
    logDebug("  New title: \(response.title ?? "cleared")")
    
    return response
  }
  
  func deleteUserData(userId: String) async throws -> DeleteUserDataResponse {
    logInfo("⚠️ Deleting ALL data for user: \(userId) (GDPR)")

    let endpoint = APIEndpoint(
      path: "/conversations/user/\(userId)",
      method: .delete
    )

    let response: DeleteUserDataResponse = try await apiClient.performRequest(endpoint)

    logInfo("Successfully deleted all user data for: \(userId)")
    logDebug("  Deleted: \(response.deleted.conversations) conversations, \(response.deleted.messages) messages, \(response.deleted.memories) memories")
    logDebug("  RAG deleted: \(response.ragDeleted)")
    
    return response
  }
}

File: android/lib/src/main/kotlin/com/mia21/services/ConversationService.kt
/**
 * ConversationService.kt
 * Mia21
 *
 * Created on November 25, 2025.
 * Copyright © 2025 Mia21. All rights reserved.
 *
 * Description:
 * Service layer for conversation history management.
 */

package com.mia21.services

import com.mia21.models.ConversationDetail
import com.mia21.models.ConversationSummary
import com.mia21.models.DeleteConversationResponse
import com.mia21.models.DeleteUserDataResponse
import com.mia21.models.RenameConversationResponse
import com.mia21.network.APIClient
import com.mia21.network.APIEndpoint
import com.mia21.network.HTTPMethod
import kotlinx.serialization.decodeFromString

/**
 * Service for managing conversation history
 */
class ConversationService(private val apiClient: APIClient) {
    
    /**
     * List conversations for a user
     */
    suspend fun listConversations(
        userId: String,
        spaceId: String? = null,
        limit: Int = 50
    ): List<ConversationSummary> {
        val queryParams = mutableMapOf(
            "user_id" to userId,
            "limit" to limit.toString()
        )
        spaceId?.let { queryParams["space_id"] = it }
        
        val path = "/conversations?" + queryParams.map { "${it.key}=${it.value}" }.joinToString("&")
        val endpoint = APIEndpoint(
            path = path,
            method = HTTPMethod.GET
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        // Parse JSON response
        return try {
            apiClient.json.decodeFromString<List<ConversationSummary>>(jsonResponse)
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * Get a specific conversation
     */
    suspend fun getConversation(conversationId: String): ConversationDetail {
        val endpoint = APIEndpoint(
            path = "/conversations/$conversationId",
            method = HTTPMethod.GET
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        return apiClient.json.decodeFromString<ConversationDetail>(jsonResponse)
    }
    
    /**
     * Delete a conversation
     */
    suspend fun deleteConversation(conversationId: String): DeleteConversationResponse {
        val endpoint = APIEndpoint(
            path = "/conversations/$conversationId",
            method = HTTPMethod.DELETE
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        return apiClient.json.decodeFromString<DeleteConversationResponse>(jsonResponse)
    }
    
    /**
     * Rename a conversation (update its title)
     * @param conversationId The conversation ID to rename
     * @param title New title for the conversation (empty string to clear)
     * @return RenameConversationResponse with success status and new title
     */
    suspend fun renameConversation(conversationId: String, title: String): RenameConversationResponse {
        val body = mapOf("title" to title)
        
        val endpoint = APIEndpoint(
            path = "/conversations/$conversationId",
            method = HTTPMethod.PATCH,
            body = body
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        return apiClient.json.decodeFromString<RenameConversationResponse>(jsonResponse)
    }
    
    /**
     * Delete ALL data for a specific end-user (GDPR compliance)
     * ⚠️ This permanently deletes all conversations, messages, memories, and RAG/vector data.
     * This action cannot be undone.
     * @param userId The end-user ID whose data should be deleted
     * @return DeleteUserDataResponse with counts of deleted items
     */
    suspend fun deleteUserData(userId: String): DeleteUserDataResponse {
        val endpoint = APIEndpoint(
            path = "/conversations/user/$userId",
            method = HTTPMethod.DELETE
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        return apiClient.json.decodeFromString<DeleteUserDataResponse>(jsonResponse)
    }
}

